client/allocrunner/consul_grpc_sock_hook.go

```diff
@@ -184,55 +185,78 @@ func (p *grpcSocketProxy) run(alloc *structs.Allocation) error {
+		if config.GRPCAddr == "" && config.Addr == "" {
+			retErr = errors.New("consul address must be set on nomad client")
+			return false
 		}
-		destAddr = net.JoinHostPort(host, p.consulGRPCFallbackPort)
-	}
 
-	hostGRPCSocketPath := filepath.Join(p.allocDir.AllocDir, allocdir.AllocGRPCSocket)
+		destAddr := config.GRPCAddr
+		if destAddr == "" {
+			// No GRPCAddr defined. Use Addr but replace port with the gRPC
+			// default of 8502.
+			host, _, err := net.SplitHostPort(config.Addr)
+			if err != nil {
+				retErr = fmt.Errorf("error parsing Consul address %q: %v",
+					config.Addr, err)
+				return false
+			}
+			destAddr = net.JoinHostPort(host, p.consulGRPCFallbackPort)
+		}
 
-	// if the socket already exists we'll try to remove it, but if not then any
-	// other errors will bubble up to the caller here or when we try to listen
-	_, err := os.Stat(hostGRPCSocketPath)
-	if err == nil {
-		err := os.Remove(hostGRPCSocketPath)
+		hostGRPCSocketPath := filepath.Join(p.allocDir.AllocDir, allocdir.AllocGRPCSocket)
+
+		// if the socket already exists we'll try to remove it, but if not then any
+		// other errors will bubble up to the caller here or when we try to listen
+		_, err := os.Stat(hostGRPCSocketPath)
+		if err == nil {
+			err := os.Remove(hostGRPCSocketPath)
+			if err != nil {
+				retErr = fmt.Errorf(
+					"unable to remove existing unix socket for Consul gRPC endpoint: %v", err)
+				return false
+			}
+		}
+
+		listener, err := net.Listen("unix", hostGRPCSocketPath)
 		if err != nil {
-			return fmt.Errorf(
-				"unable to remove existing unix socket for Consul gRPC endpoint: %v", err)
+			retErr = fmt.Errorf("unable to create unix socket for Consul gRPC endpoint: %v", err)
+			return false
 		}
-	}
 
-	listener, err := net.Listen("unix", hostGRPCSocketPath)
-	if err != nil {
-		return fmt.Errorf("unable to create unix socket for Consul gRPC endpoint: %v", err)
-	}
+		// The gRPC socket should be usable by all users in case a task is
+		// running as an unprivileged user.  Unix does not allow setting domain
+		// socket permissions when creating the file, so we must manually call
+		// chmod afterwards.
+		// https://github.com/golang/go/issues/11822
+		if err = os.Chmod(hostGRPCSocketPath, os.ModePerm); err != nil {
+			retErr = fmt.Errorf("unable to set permissions on unix socket for Consul gRPC endpoint: %v", err)
+			return false
+		}
 
-	// The gRPC socket should be usable by all users in case a task is
-	// running as an unprivileged user.  Unix does not allow setting domain
-	// socket permissions when creating the file, so we must manually call
-	// chmod afterwards.
-	// https://github.com/golang/go/issues/11822
-	if err := os.Chmod(hostGRPCSocketPath, os.ModePerm); err != nil {
-		return fmt.Errorf("unable to set permissions on unix socket for Consul gRPC endpoint: %v", err)
-	}
+		go func() {
+			proxy(p.ctx, p.logger, destAddr, listener)
+			p.cancel()
+			close(p.doneCh)
+		}()
```

Comment:
Hum...this func makes me think we're iterating at the wrong level 🤔 

From what I understood from this part of the code, a `grpcSocketProxy` is used to expose Consul's gRPC endpoint to the task. By iterating over clusters here we would be overwriting `hostGRPCSocketPath` so that only the last cluster is being reached over this socket.

And, when the `proxy()` function returns here we will `close(p.doneCh)` multiple times, causing a `panic`.

I feel like the correct behaviour would be for `consulGRPCSocketHook` to hold a map of proxies, one for each cluster used in the alloc. Then each proxy is started pointing to a single Consul config.

Reply:
Oof, good catch.

Reply:
Fixed!